In [ ]:
words = open('names.txt', 'r').read().splitlines()

In [ ]:
chars = sorted(list(set(''.join(words))))

In [ ]:
new_chars = ['.'] + chars
list_twos = []
for ch1 in new_chars:
  for ch2 in new_chars[1:]:
    list_twos.append(ch1+ch2)
list_twos.append('..')
pairs = sorted(list_twos)
len(pairs)
# pairs

703

In [ ]:
stoi_2 = {s:i for i, s in enumerate(pairs)}
itos_2 = {i:s for s, i in stoi_2.items()}
stoi_1 = {s:i for i, s in enumerate(new_chars)}
itos_1 = {i:s for s, i in stoi_1.items()}

In [ ]:
# Counting method for trigram model

In [ ]:
import torch

In [ ]:
N = torch.zeros((703, 27), dtype = torch.int32)

In [ ]:
for w in words:
  chs = ['.'] + ['.'] + list(w) + ['.']
  for x, y, z in zip(chs, chs[1:], chs[2:]):
    xs = (x+y)
    ys = z
    # print(xs, ys)
    indx1 = stoi_2[xs]
    indx2 = stoi_1[ys]
    N[indx1, indx2]+=1

In [ ]:
P = (N+1).float()
P/=P.sum(1, keepdims=True)

In [ ]:
log_likelihood = 0.0
n = 0

for w in words:
  chs = ['.'] + ['.'] + list(w) + ['.']
  for x, y, z in zip(chs, chs[1:], chs[2:]):
    xs = (x+y)
    ys = z
    # print(xs, ys)
    indx1 = stoi_2[xs]
    indx2 = stoi_1[ys]
    prob = P[indx1, indx2]
    logprob = torch.log(prob)
    log_likelihood += logprob
    n += 1

nll = -log_likelihood
print(f'{nll/n}')

2.2119739055633545


In [ ]:
# Counting method for bigram model

In [ ]:
Nb = torch.zeros((27, 27), dtype = torch.int32)

In [ ]:
for w in words:
  chs = ['.'] + list(w) + ['.']
  for ch1, ch2 in zip(ch2, ch2[1:]):
    indx1 = stoi_1[ch1]
    indx2 = stoi_1[ch2]
    Nb[indx1, indx2] += 1

In [ ]:
Pb = (Nb+1).float()
Pb/=Pb.sum(1, keepdims=True)

In [ ]:
log_likelihood = 0.0
n = 0

for w in words:
  chs = ['.'] + list(w) + ['.']
  for ch1, ch2 in zip(chs, chs[1:]):
    indx1 = stoi_1[ch1]
    indx2 = stoi_1[ch2]
    prob = Pb[indx1, indx2]
    logprob = torch.log(prob)
    log_likelihood += logprob
    n += 1

nll = -log_likelihood
print(f'{nll/n}')

3.2960469722747803


**Observe that the loss in case of trigram model (2.21..) is less than the loss in case of the bigram model (3.29..)**

In [ ]:
# Counting method of trigram with train, dev and test sets

In [ ]:
import random
random.seed(42)
random.shuffle(words)
n1 = int(0.8*len(words))
n2 = int(0.9*len(words))

train_set = words[:n1]
dev_set = words[n1:n2]
test_set = words[n2:]

In [ ]:
N = torch.zeros((703, 27), dtype = torch.int32)

In [ ]:
for w in train_set:
  chs = ['.'] + ['.'] + list(w) + ['.']
  for x, y, z in zip(chs, chs[1:], chs[2:]):
    xs = (x+y)
    ys = z
    # print(xs, ys)
    indx1 = stoi_2[xs]
    indx2 = stoi_1[ys]
    N[indx1, indx2]+=1

In [ ]:
P = (N+1).float()
P/=P.sum(1, keepdims=True)

In [ ]:
log_likelihood = 0.0
n = 0

for w in train_set:
  chs = ['.'] + ['.'] + list(w) + ['.']
  for x, y, z in zip(chs, chs[1:], chs[2:]):
    xs = (x+y)
    ys = z
    # print(xs, ys)
    indx1 = stoi_2[xs]
    indx2 = stoi_1[ys]
    prob = P[indx1, indx2]
    logprob = torch.log(prob)
    log_likelihood += logprob
    n += 1

nll = -log_likelihood
print(f'train set loss: {nll/n}')

train set loss: 2.2157092094421387


In [ ]:
log_likelihood = 0.0
n = 0

for w in dev_set:
  chs = ['.'] + ['.'] + list(w) + ['.']
  for x, y, z in zip(chs, chs[1:], chs[2:]):
    xs = (x+y)
    ys = z
    # print(xs, ys)
    indx1 = stoi_2[xs]
    indx2 = stoi_1[ys]
    prob = P[indx1, indx2]
    logprob = torch.log(prob)
    log_likelihood += logprob
    n += 1

nll = -log_likelihood
print(f'dev set loss: {nll/n}')

dev set loss: 2.236459732055664


In [ ]:
log_likelihood = 0.0
n = 0

for w in test_set:
  chs = ['.'] + ['.'] + list(w) + ['.']
  for x, y, z in zip(chs, chs[1:], chs[2:]):
    xs = (x+y)
    ys = z
    # print(xs, ys)
    indx1 = stoi_2[xs]
    indx2 = stoi_1[ys]
    prob = P[indx1, indx2]
    logprob = torch.log(prob)
    log_likelihood += logprob
    n += 1

nll = -log_likelihood
print(f'test set loss {nll/n}')

test set loss 2.2373220920562744


In [ ]:
# Counting method of bigram with train, dev and test sets

In [ ]:
Nb = torch.zeros((27, 27), dtype = torch.int32)

In [ ]:
for w in train_set:
  chs = ['.'] + list(w) + ['.']
  for ch1, ch2 in zip(ch2, ch2[1:]):
    indx1 = stoi_1[ch1]
    indx2 = stoi_1[ch2]
    Nb[indx1, indx2] += 1

In [ ]:
Pb = (Nb+1).float()
Pb/=Pb.sum(1, keepdims=True)

In [ ]:
log_likelihood = 0.0
n = 0

for w in train_set:
  chs = ['.'] + ['.'] + list(w) + ['.']
  for x, y, z in zip(chs, chs[1:], chs[2:]):
    xs = (x+y)
    ys = z
    # print(xs, ys)
    indx1 = stoi_2[xs]
    indx2 = stoi_1[ys]
    prob = P[indx1, indx2]
    logprob = torch.log(prob)
    log_likelihood += logprob
    n += 1

nll = -log_likelihood
print(f'train set loss: {nll/n}')

train set loss: 2.2157092094421387


In [ ]:
log_likelihood = 0.0
n = 0

for w in dev_set:
  chs = ['.'] + list(w) + ['.']
  for ch1, ch2 in zip(chs, chs[1:]):
    indx1 = stoi_1[ch1]
    indx2 = stoi_1[ch2]
    prob = Pb[indx1, indx2]
    logprob = torch.log(prob)
    log_likelihood += logprob
    n += 1

nll = -log_likelihood
print(f'dev set loss {nll/n}')

dev set loss 3.296229362487793


In [ ]:
log_likelihood = 0.0
n = 0

for w in test_set:
  chs = ['.'] + list(w) + ['.']
  for ch1, ch2 in zip(chs, chs[1:]):
    indx1 = stoi_1[ch1]
    indx2 = stoi_1[ch2]
    prob = Pb[indx1, indx2]
    logprob = torch.log(prob)
    log_likelihood += logprob
    n += 1

nll = -log_likelihood
print(f'test set loss {nll/n}')

test set loss 3.2962353229522705


**Observing that the train_set loss and dev or test set losses does not differ much we can say that the model has not overfitted.**


In [ ]:
# Smoothing on trigram model

In [ ]:
N = torch.zeros((703, 27), dtype = torch.int32)

for w in train_set:
  chs = ['.'] + ['.'] + list(w) + ['.']
  for x, y, z in zip(chs, chs[1:], chs[2:]):
    xs = (x+y)
    ys = z
    # print(xs, ys)
    indx1 = stoi_2[xs]
    indx2 = stoi_1[ys]
    N[indx1, indx2]+=1

P = (N+1).float()
# P = (N+10).float()
# P = (N+100).float()
P/=P.sum(1, keepdims=True)

log_likelihood = 0.0
n = 0

for w in train_set:
  chs = ['.'] + ['.'] + list(w) + ['.']
  for x, y, z in zip(chs, chs[1:], chs[2:]):
    xs = (x+y)
    ys = z
    # print(xs, ys)
    indx1 = stoi_2[xs]
    indx2 = stoi_1[ys]
    prob = P[indx1, indx2]
    logprob = torch.log(prob)
    log_likelihood += logprob
    n += 1

nll = -log_likelihood
print(f'train set loss: {nll/n}')

log_likelihood = 0.0
n = 0

for w in dev_set:
  chs = ['.'] + ['.'] + list(w) + ['.']
  for x, y, z in zip(chs, chs[1:], chs[2:]):
    xs = (x+y)
    ys = z
    # print(xs, ys)
    indx1 = stoi_2[xs]
    indx2 = stoi_1[ys]
    prob = P[indx1, indx2]
    logprob = torch.log(prob)
    log_likelihood += logprob
    n += 1

nll = -log_likelihood
print(f'dev set loss: {nll/n}')

train set loss: 2.2157092094421387
dev set loss: 2.236459732055664


In [ ]:
N = torch.zeros((703, 27), dtype = torch.int32)

for w in train_set:
  chs = ['.'] + ['.'] + list(w) + ['.']
  for x, y, z in zip(chs, chs[1:], chs[2:]):
    xs = (x+y)
    ys = z
    # print(xs, ys)
    indx1 = stoi_2[xs]
    indx2 = stoi_1[ys]
    N[indx1, indx2]+=1

# P = (N+1).float()
P = (N+10).float()
# P = (N+100).float()
P/=P.sum(1, keepdims=True)

log_likelihood = 0.0
n = 0

for w in train_set:
  chs = ['.'] + ['.'] + list(w) + ['.']
  for x, y, z in zip(chs, chs[1:], chs[2:]):
    xs = (x+y)
    ys = z
    # print(xs, ys)
    indx1 = stoi_2[xs]
    indx2 = stoi_1[ys]
    prob = P[indx1, indx2]
    logprob = torch.log(prob)
    log_likelihood += logprob
    n += 1

nll = -log_likelihood
print(f'train set loss: {nll/n}')

log_likelihood = 0.0
n = 0

for w in dev_set:
  chs = ['.'] + ['.'] + list(w) + ['.']
  for x, y, z in zip(chs, chs[1:], chs[2:]):
    xs = (x+y)
    ys = z
    # print(xs, ys)
    indx1 = stoi_2[xs]
    indx2 = stoi_1[ys]
    prob = P[indx1, indx2]
    logprob = torch.log(prob)
    log_likelihood += logprob
    n += 1

nll = -log_likelihood
print(f'dev set loss: {nll/n}')

train set loss: 2.356320381164551
dev set loss: 2.363665819168091


In [ ]:
N = torch.zeros((703, 27), dtype = torch.int32)

for w in train_set:
  chs = ['.'] + ['.'] + list(w) + ['.']
  for x, y, z in zip(chs, chs[1:], chs[2:]):
    xs = (x+y)
    ys = z
    # print(xs, ys)
    indx1 = stoi_2[xs]
    indx2 = stoi_1[ys]
    N[indx1, indx2]+=1

# P = (N+1).float()
# P = (N+10).float()
P = (N+100).float()
P/=P.sum(1, keepdims=True)

log_likelihood = 0.0
n = 0

for w in train_set:
  chs = ['.'] + ['.'] + list(w) + ['.']
  for x, y, z in zip(chs, chs[1:], chs[2:]):
    xs = (x+y)
    ys = z
    # print(xs, ys)
    indx1 = stoi_2[xs]
    indx2 = stoi_1[ys]
    prob = P[indx1, indx2]
    logprob = torch.log(prob)
    log_likelihood += logprob
    n += 1

nll = -log_likelihood
print(f'train set loss: {nll/n}')

log_likelihood = 0.0
n = 0

for w in dev_set:
  chs = ['.'] + ['.'] + list(w) + ['.']
  for x, y, z in zip(chs, chs[1:], chs[2:]):
    xs = (x+y)
    ys = z
    # print(xs, ys)
    indx1 = stoi_2[xs]
    indx2 = stoi_1[ys]
    prob = P[indx1, indx2]
    logprob = torch.log(prob)
    log_likelihood += logprob
    n += 1

nll = -log_likelihood
print(f'dev set loss: {nll/n}')

train set loss: 2.7453043460845947
dev set loss: 2.746373176574707


In [ ]:
N = torch.zeros((703, 27), dtype = torch.int32)

for w in train_set:
  chs = ['.'] + ['.'] + list(w) + ['.']
  for x, y, z in zip(chs, chs[1:], chs[2:]):
    xs = (x+y)
    ys = z
    # print(xs, ys)
    indx1 = stoi_2[xs]
    indx2 = stoi_1[ys]
    N[indx1, indx2]+=1

# P = (N+1).float()
# P = (N+10).float()
# P = (N+100).float()
P = (N+1000).float()
P/=P.sum(1, keepdims=True)

log_likelihood = 0.0
n = 0

for w in train_set:
  chs = ['.'] + ['.'] + list(w) + ['.']
  for x, y, z in zip(chs, chs[1:], chs[2:]):
    xs = (x+y)
    ys = z
    # print(xs, ys)
    indx1 = stoi_2[xs]
    indx2 = stoi_1[ys]
    prob = P[indx1, indx2]
    logprob = torch.log(prob)
    log_likelihood += logprob
    n += 1

nll = -log_likelihood
print(f'train set loss: {nll/n}')

log_likelihood = 0.0
n = 0

for w in dev_set:
  chs = ['.'] + ['.'] + list(w) + ['.']
  for x, y, z in zip(chs, chs[1:], chs[2:]):
    xs = (x+y)
    ys = z
    # print(xs, ys)
    indx1 = stoi_2[xs]
    indx2 = stoi_1[ys]
    prob = P[indx1, indx2]
    logprob = torch.log(prob)
    log_likelihood += logprob
    n += 1

nll = -log_likelihood
print(f'dev set loss: {nll/n}')

train set loss: 3.1363141536712646
dev set loss: 3.1366119384765625


**Observe that the train set loss and dev set loss are increasing as the value used for smoothing is increased. So the best setting value is 1 i.e., P = (N+1).float().**

In [ ]:
N = torch.zeros((703, 27), dtype = torch.int32)

for w in train_set:
  chs = ['.'] + ['.'] + list(w) + ['.']
  for x, y, z in zip(chs, chs[1:], chs[2:]):
    xs = (x+y)
    ys = z
    # print(xs, ys)
    indx1 = stoi_2[xs]
    indx2 = stoi_1[ys]
    N[indx1, indx2]+=1

P = (N+1).float()
# P = (N+10).float()
# P = (N+100).float()
P/=P.sum(1, keepdims=True)

log_likelihood = 0.0
n = 0

for w in train_set:
  chs = ['.'] + ['.'] + list(w) + ['.']
  for x, y, z in zip(chs, chs[1:], chs[2:]):
    xs = (x+y)
    ys = z
    # print(xs, ys)
    indx1 = stoi_2[xs]
    indx2 = stoi_1[ys]
    prob = P[indx1, indx2]
    logprob = torch.log(prob)
    log_likelihood += logprob
    n += 1

nll = -log_likelihood
print(f'train set loss: {nll/n}')

log_likelihood = 0.0
n = 0

for w in dev_set:
  chs = ['.'] + ['.'] + list(w) + ['.']
  for x, y, z in zip(chs, chs[1:], chs[2:]):
    xs = (x+y)
    ys = z
    # print(xs, ys)
    indx1 = stoi_2[xs]
    indx2 = stoi_1[ys]
    prob = P[indx1, indx2]
    logprob = torch.log(prob)
    log_likelihood += logprob
    n += 1

nll = -log_likelihood
print(f'dev set loss: {nll/n}')

log_likelihood = 0.0
n = 0

for w in test_set:
  chs = ['.'] + ['.'] + list(w) + ['.']
  for x, y, z in zip(chs, chs[1:], chs[2:]):
    xs = (x+y)
    ys = z
    # print(xs, ys)
    indx1 = stoi_2[xs]
    indx2 = stoi_1[ys]
    prob = P[indx1, indx2]
    logprob = torch.log(prob)
    log_likelihood += logprob
    n += 1

nll = -log_likelihood
print(f'test set loss: {nll/n}')

train set loss: 2.2157092094421387
dev set loss: 2.236459732055664
test set loss: 2.2373220920562744


In [ ]:
g = torch.Generator().manual_seed(2147483647)
W = torch.randn((27, 27), generator=g, requires_grad=True)

In [ ]:
xs, ys = [], []

for w in words:
    # print(w)
    chs = ['.'] + list(w) + ['.']
    for ch1, ch2 in zip(chs, chs[1:]):
        indx1 = stoi_1[ch1]
        indx2 = stoi_1[ch2]
        xs.append(indx1)
        ys.append(indx2)

xs = torch.tensor(xs)
ys = torch.tensor(ys)

In [ ]:
xs

tensor([ 0,  5, 13,  ..., 25, 26, 24])

**We can directly index the indices of the which we need from W into W i.e., multiplying one hot encoded vector of index 0 is same as indexing 0 into W.**

In [ ]:
import torch.nn.functional as F
temp = torch.tensor([0])
enc_temp = F.one_hot(temp, num_classes=27).float()
print(enc_temp @ W)
print(W[temp])

tensor([[ 1.5674, -0.2373, -0.0274, -1.1008,  0.2859, -0.0296, -1.5471,  0.6049,
          0.0791,  0.9046, -0.4713,  0.7868, -0.3284, -0.4330,  1.3729,  2.9334,
          1.5618, -1.6261,  0.6772, -0.8404,  0.9849, -0.1484, -1.4795,  0.4483,
         -0.0707,  2.4968,  2.4448]], grad_fn=<MmBackward0>)
tensor([[ 1.5674, -0.2373, -0.0274, -1.1008,  0.2859, -0.0296, -1.5471,  0.6049,
          0.0791,  0.9046, -0.4713,  0.7868, -0.3284, -0.4330,  1.3729,  2.9334,
          1.5618, -1.6261,  0.6772, -0.8404,  0.9849, -0.1484, -1.4795,  0.4483,
         -0.0707,  2.4968,  2.4448]], grad_fn=<IndexBackward0>)


In [ ]:
W[xs]

tensor([[ 1.5674, -0.2373, -0.0274,  ..., -0.0707,  2.4968,  2.4448],
        [ 0.4724,  1.4830,  0.3175,  ..., -0.4275, -2.1259,  0.9604],
        [ 0.1936,  1.0532,  0.6339,  ...,  1.5447,  0.6006, -0.7091],
        ...,
        [ 0.7414, -0.5879, -0.4651,  ..., -0.1388,  1.3096, -0.2580],
        [ 1.0669,  0.2136, -0.7660,  ...,  1.3405, -0.2175,  0.8627],
        [ 0.2163, -0.7873, -0.3301,  ..., -0.2630, -0.7552,  0.8191]],
       grad_fn=<IndexBackward0>)

In [ ]:
enc = F.one_hot(xs, num_classes=27).float()

In [ ]:
enc @ W

tensor([[ 1.5674, -0.2373, -0.0274,  ..., -0.0707,  2.4968,  2.4448],
        [ 0.4724,  1.4830,  0.3175,  ..., -0.4275, -2.1259,  0.9604],
        [ 0.1936,  1.0532,  0.6339,  ...,  1.5447,  0.6006, -0.7091],
        ...,
        [ 0.7414, -0.5879, -0.4651,  ..., -0.1388,  1.3096, -0.2580],
        [ 1.0669,  0.2136, -0.7660,  ...,  1.3405, -0.2175,  0.8627],
        [ 0.2163, -0.7873, -0.3301,  ..., -0.2630, -0.7552,  0.8191]],
       grad_fn=<MmBackward0>)

In [ ]:
enc @ W == W[xs]

tensor([[True, True, True,  ..., True, True, True],
        [True, True, True,  ..., True, True, True],
        [True, True, True,  ..., True, True, True],
        ...,
        [True, True, True,  ..., True, True, True],
        [True, True, True,  ..., True, True, True],
        [True, True, True,  ..., True, True, True]])

In [ ]:
xs, ys = [], []
N = torch.zeros((27, 27), dtype = torch.int32)

for w in words:
    # print(w)
    chs = ['.'] + list(w) + ['.']
    for ch1, ch2 in zip(chs, chs[1:]):
        indx1 = stoi_1[ch1]
        indx2 = stoi_1[ch2]
        N[indx1, indx2] += 1
        xs.append(indx1)
        ys.append(indx2)

xs = torch.tensor(xs)
ys = torch.tensor(ys)

In [ ]:
P = (N+1).float()
P /= P.sum(1, keepdims=True)

In [ ]:
logits = P[xs].log()
logits

tensor([[-10.3754,  -1.9835,  -3.1999,  ...,  -5.4701,  -4.0912,  -3.5402],
        [ -1.6357,  -3.4036,  -5.1217,  ...,  -5.0354,  -2.9494,  -4.7217],
        [ -2.5572,  -0.9454,  -4.0778,  ...,  -8.8052,  -3.1423,  -6.3203],
        ...,
        [ -1.5855,  -1.5200,  -5.8582,  ...,  -5.8231,  -6.0124,  -4.8210],
        [ -2.7122,  -1.0355,  -6.1841,  ...,  -7.1004,  -2.7964,  -3.9649],
        [ -1.4788,  -1.9404,  -5.8916,  ...,  -2.9212,  -3.1508,  -3.5891]])

In [ ]:
loss = F.cross_entropy(logits, ys)
loss

tensor(2.4546)

In [ ]:
log_likelihoods = []

for w in words:
  chs = ['.'] + list(w) + ['.']
  for ch1, ch2 in zip(chs, chs[1:]):
    # print(xs, ys)
    indx1 = stoi_1[ch1]
    indx2 = stoi_1[ch2]
    prob = P[indx1, indx2]
    logprob = torch.log(prob)
    log_likelihoods.append(logprob)

logs = torch.tensor(log_likelihoods)
print(f'test set loss: {-logs.mean():.4f}')

test set loss: 2.4546


**Clearly the loss obtained by calculating negative log likelihoods is same as the loss obtained by using cross_entropy.**

In [ ]:
logits = torch.tensor([-5, 0, 3, 100])
counts = logits.exp()
probs = counts/counts.sum()
probs

tensor([0., 0., 0., nan])

**While cross_entropy function substracts the maximum value of the logits and calculates the probabilities which does not allow the probabilies to become nan**

In [ ]:
logits = torch.tensor([-5, 0, 3, 5])
counts = logits.exp()
probs = counts/counts.sum()
probs

tensor([3.9751e-05, 5.8995e-03, 1.1849e-01, 8.7557e-01])

In [ ]:
logits = torch.tensor([-5, 0, 3, 5]) - 5
counts = logits.exp()
probs = counts/counts.sum()
probs

tensor([3.9751e-05, 5.8995e-03, 1.1849e-01, 8.7557e-01])

**Observe that the probabilities does not change by doing so (subtracting the maximum value).**

**Conclusion: cross_entropy function is numerically more behaved than direct calculation**

**Meta-Exercise - Generating words using the trigram model we trained**

In [8]:
# Once again the trigram model

In [1]:
words = open('names.txt', 'r').read().splitlines()

In [2]:
chars = sorted(list(set(''.join(words))))

In [3]:
new_chars = ['.'] + chars
list_twos = []
for ch1 in new_chars:
  for ch2 in new_chars[1:]:
    list_twos.append(ch1+ch2)
list_twos.append('..')
pairs = sorted(list_twos)
len(pairs)
# pairs

703

In [4]:
stoi_2 = {s:i for i, s in enumerate(pairs)}
itos_2 = {i:s for s, i in stoi_2.items()}
stoi_1 = {s:i for i, s in enumerate(new_chars)}
itos_1 = {i:s for s, i in stoi_1.items()}

In [5]:
import torch

In [10]:
N = torch.zeros((703, 27), dtype = torch.int32)

In [11]:
for w in words:
  chs = ['.'] + ['.'] + list(w) + ['.']
  for x, y, z in zip(chs, chs[1:], chs[2:]):
    xs = (x+y)
    ys = z
    # print(xs, ys)
    indx1 = stoi_2[xs]
    indx2 = stoi_1[ys]
    N[indx1, indx2]+=1

In [12]:
P = (N+1).float()
P/=P.sum(1, keepdims=True)

In [13]:
log_likelihood = 0.0
n = 0

for w in words:
  chs = ['.'] + ['.'] + list(w) + ['.']
  for x, y, z in zip(chs, chs[1:], chs[2:]):
    xs = (x+y)
    ys = z
    # print(xs, ys)
    indx1 = stoi_2[xs]
    indx2 = stoi_1[ys]
    prob = P[indx1, indx2]
    logprob = torch.log(prob)
    log_likelihood += logprob
    n += 1

nll = -log_likelihood
print(f'{nll/n}')

2.2119739055633545


In [15]:
g = torch.Generator().manual_seed(2147483647+10)

for _ in range(30):
    out = []
    temp = [0, 0]
    while True:
        chs = itos_1[temp[0]] + itos_1[temp[1]]
        indx = stoi_2[chs]
        p = P[indx]
        indx_next = torch.multinomial(p, num_samples=1, replacement=True, generator=g).item()
        if indx_next==0:
            break
        temp = temp[1:] + [indx_next]
        out.append(itos_1[indx_next])

    print(''.join(out))

mon
ammyan
see
madhriah
rethruthadraeg
aderedielin
shi
jen
edelieananarleilyn
hona
ca
shubvrgihimiel
kin
renelilanteromius
kavder
yahleyeh
yumajaystondrihil
salynnsuf
zakel
june
jart
kiveaosten
adi
fen
oloeh
zptin
samuezwhkhareon
is
kiri
evon
